Chess Piece tests

frozenfrank opened the pull request:
## Key Points
- **This PR does _not_ change any requirements.** It refactors duplication out of the tests.
- This PR does make explicit an otherwise implicit requirement: that `ChessBoard` have be able to pass `equals()` and `hashCode()` testing. Previously, these tests were omitted, but a student couldn't pass the course without providing the functionality.
- These improved tests would have clarified a tricky condition where a student was passing all tests locally, but not on the autograder. The same code on these expanded tests failed locally.

## Overview

Add new `ChessPieceTests` which highlight the need for students to provide `equals()` and `hashCode()` methods for the class. This also refactors equals/hashcode testing generally to remove duplicate and rely on a shared code base. The addition of these tests to `ChessBoard` came nearly for free.

## Discussion

This PR moves & changes the "Piece Moves on All Pieces" into the `ChessPieceTests` file.

See the [screenshots in the comments](https://github.com/softwareconstruction240/chess/pull/16#issuecomment-2639024147) about how this solved a tricky problem for a student.

Resolves #11.

leesjensen reviewed (approved):
Thanks!

frozenfrank committed:
Rename chess/piece ==> chess/piecemoves tests... Resolves #15

frozenfrank committed:
Create ChessPieceTests with copied equals methods

frozenfrank committed:
Rename combined equals & hashcode testing for clarity

frozenfrank committed:
Introduce EqualsTestingUtility and install in ChessPieceTests

frozenfrank committed:
Simplify and improve EqualsTestingUtility interface

frozenfrank committed:
Add thoroughness to ChessPieceTests allDifferent

frozenfrank committed:
Improve debugability of EqualsTestingUtility...

- Add class comment describing role of class
- Require className to be provided as context
- Use className to make error messages more specific
- Add error messages to all remaining assertions

frozenfrank committed:
Install EqualsTestingUtility in ChessMoveTests

frozenfrank committed:
Install EqualsTestingUtility in ChessPositionTests

frozenfrank committed:
Install EqualsTestingUtility in ChessBoardTests

frozenfrank committed:
Move 'Piece Move on All Pieces' to ChessPieceTests...

In order to break the dependency on ChessBoard.resetBoard(),
this also switches the test to simply evaluating pieceMoves()
on all possible piece types on all possible squares on the board.

768 total evaluations is a high number, but it also gives us
strong confidence that the code never throws errors!

Maillman reviewed (approved):
Looks good.

ZakkeryDaRebel commented on shared/src/test/java/passoff/chess/ChessMoveTests.java:
```diff
@@ -3,77 +3,35 @@
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.ROOK),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.BISHOP),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.KNIGHT),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.PAWN)
```
Are all of these new ChessMoves indented one too far?

frozenfrank replied:
This the way Intelli-J did it by default. It's new to me as well (from other languages/IDE's), but Intelli-J liks to have a long intent for these kinds of things.

frozenfrank replied:
I'm going to close this as resolved since it's just the default behavior.

Notice that in the KnightMoveTests, they also have a long hanging indent, plus extra indentation for the string alignment of the `boardText`:

https://github.com/softwareconstruction240/chess/blob/b3d94c165012348ecf7abf2a7abd71479e3ce7b7/shared/src/test/java/passoff/chess/piece/KnightMoveTests.java#L28-L45

Here is another example of this long hanging indent from the autograder project:

https://github.com/softwareconstruction240/autograder/blob/60528d0751ab38842bec6a35c4f29fe9ebb7eb05/src/test/java/edu/byu/cs/autograder/score/ScorerTest.java#L262-L270

ZakkeryDaRebel commented on shared/src/test/java/passoff/chess/ChessPieceTests.java:
```diff
@@ -0,0 +1,67 @@
+                new ChessPiece(ChessGame.TeamColor.BLACK, ChessPiece.PieceType.QUEEN),
+                new ChessPiece(ChessGame.TeamColor.WHITE, ChessPiece.PieceType.PAWN),
+                new ChessPiece(ChessGame.TeamColor.BLACK, ChessPiece.PieceType.PAWN)
+        );
```
Same indention question as above

frozenfrank replied:
https://github.com/softwareconstruction240/chess/pull/16#discussion_r1932921478

ZakkeryDaRebel reviewed (approved):
Besides the indentation questions, this all looks good to me. Smart to combine the similar equals and hashcode testing into one location.

frozenfrank commented:
> Did you mean to title this about position? I thought this was supposed to be about piece

@19mdavenport  Yes, you are right. I did mean to title this about `ChessPieceTests`. Thanks for pointing that out.

Unfortunately, I can't rename the branch, but I can rename and edit everything else...

19mdavenport commented on shared/src/test/java/passoff/chess/ChessPieceTests.java:
```diff
@@ -0,0 +1,67 @@
+                .map(teamColor -> new ChessPiece(teamColor, pieceType)))
+                .toList();
+
+        // 8 rows * 8 cols * 12 pieces = 768 evaluations
```
Is it valid to have a white pawn on row 8 or a black pawn on row 1? My implementation assumes this isn't valid, but I can change it if necessary

frozenfrank replied:
We already hashed this out thoroughly in person. Consistent with the motivation behind 38c4f14487b483ee9ac922cc124babc63576fe03, these assumption challenging decisions have been removed from the code.

19mdavenport commented on shared/src/test/java/passoff/chess/ChessMoveTests.java:
```diff
@@ -3,77 +3,35 @@
+                new ChessMove(new ChessPosition(1, 5), new ChessPosition(2, 6), null),
+                new ChessMove(new ChessPosition(2, 4), new ChessPosition(1, 5), null),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(5, 3), null),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.KING), // hehe :)
```
Bruh. You really wanna penalize people for extra error checking?

19mdavenport commented on shared/src/test/java/passoff/chess/ChessMoveTests.java:
```diff
@@ -3,77 +3,35 @@
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.ROOK),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.BISHOP),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.KNIGHT),
+                new ChessMove(new ChessPosition(2, 6), new ChessPosition(1, 5), ChessPiece.PieceType.PAWN)
```
Bruh pt 2

19mdavenport commented on shared/src/test/java/passoff/chess/EqualsTestingUtility.java:
```diff
@@ -0,0 +1,97 @@
+ * @param <T> The type to be compared during testing.
+ */
+public abstract class EqualsTestingUtility<T> {
+    private final String className;
```
Is having className necessary? Is the name of the test class (ChessPieceTests, ChesMoveTests, etc.) not sufficient for this?

frozenfrank replied:
Yes. I want to present helpful error messages in the output.

19mdavenport reviewed (commented):
I'm concerned this may be two advanced for some beginner students to be able to debug (I know, total hypocrite, WebSocketTests.java). I guess the TestUtilities already has some of that, so maybe not a super big deal.

I like the extra testability of equality of ChessBoard, that should be a good addition for students not being confused by the resetBoard test

There's a lot of multiple String concatenation here. I've been warned against that before, although I'm not sure if there's a perfect alternative. Are you aware of the warnings against multiple string concatenation?

frozenfrank committed:
Avoid creating impossible chess scenarios...

- Prevent placing pawns on the end rows
- Remove pawn promotions to kings & pawns

While I personally think that lower-level classes should not be making
these kinds of assumptions, I also accept that students sometimes do.

Consistent with the change in 38c4f14487b483ee9ac922cc124babc63576fe03,
we want to allow students to make more assumptions than they should.

frozenfrank commented:
## Student Application

We had a student today with a very tricky problem. His tests all passed locally, and he hadn't modified the test files, but they consistently failed on the AutoGrader. Four different TA's (including myself) tried looking at it, but we were all stumped.

A fifth TA was finally able to point the student in the right direction. The two problems (missing ChessPiece equality and `static` ChessBoard attributes) would both have been caught by these improved tests.

**Student Situation**

![image](https://github.com/user-attachments/assets/dc4d4e96-3078-4059-bcda-d98cea3edc86)

**All Current Tests Passing**

![image](https://github.com/user-attachments/assets/a6fa5e23-17e6-4a98-9019-b987c502409f)

**Issue Highlighted by New Tests**

![image](https://github.com/user-attachments/assets/65b80e02-16d0-4653-8333-3c7db1bc0264)

frozenfrank commented:
@jerodw This PR is ready for review :)
